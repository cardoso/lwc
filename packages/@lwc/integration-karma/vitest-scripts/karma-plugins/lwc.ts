/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import path from 'path';
import rollupPluginLwc from '@lwc/rollup-plugin';
import type { Plugin } from 'vitest/config';

export type VitestLwcOptions = {
    dir: string;
};

export default function vitestPluginLwc(pluginOptions: VitestLwcOptions): Plugin {
    const rollupPlugin = rollupPluginLwc({
        rootDir: pluginOptions.dir,
        include: ['test/**/*.spec.js', 'test/**/*.js', 'test/**/*.html', 'test/**/*.css'],
    });

    return {
        name: 'vitest-plugin-lwc',
        enforce: 'post',

        buildStart(options) {
            // @ts-expect-error rollupPlugin is not defined
            return rollupPlugin.buildStart!.call(this, options);
        },
        resolveId(source, importer, _options) {
            if (!importer) {
                return;
            }

            if (importer.endsWith('.spec.js')) {
                rollupPlugin.api.updateOptions({
                    rootDir: path.parse(importer).dir,
                });
            }

            // @ts-expect-error rollupPlugin is not defined
            return rollupPlugin.resolveId!.call(this, source, importer);
        },
        load(id, _options) {
            // @ts-expect-error rollupPlugin is not defined
            return rollupPlugin.load!.call(this, id);
        },
        transform(code, id, _options) {
            // @ts-expect-error rollupPlugin is not defined
            return rollupPlugin.transform!.call(this, code, id);
        },
    };
}
