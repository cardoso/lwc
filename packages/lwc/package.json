{
    "name": "lwc",
    "version": "8.9.0",
    "description": "Lightning Web Components (LWC)",
    "homepage": "https://lwc.dev/",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/lwc.git",
        "directory": "packages/lwc"
    },
    "bugs": {
        "url": "https://github.com/salesforce/lwc/issues"
    },
    "main": "index.js",
    "type": "module",
    "typings": "index.d.ts",
    "license": "MIT",
    "files": [
        "*.js",
        "*.d.ts",
        "!vitest.config.*"
    ],
    "dependencies": {
        "@lwc/aria-reflection": "workspace:*",
        "@lwc/babel-plugin-component": "workspace:*",
        "@lwc/compiler": "workspace:*",
        "@lwc/engine-core": "workspace:*",
        "@lwc/engine-dom": "workspace:*",
        "@lwc/engine-server": "workspace:*",
        "@lwc/errors": "workspace:*",
        "@lwc/features": "workspace:*",
        "@lwc/module-resolver": "workspace:*",
        "@lwc/rollup-plugin": "workspace:*",
        "@lwc/shared": "workspace:*",
        "@lwc/ssr-compiler": "workspace:*",
        "@lwc/ssr-runtime": "workspace:*",
        "@lwc/style-compiler": "workspace:*",
        "@lwc/synthetic-shadow": "workspace:*",
        "@lwc/template-compiler": "workspace:*",
        "@lwc/types": "workspace:*",
        "@lwc/wire-service": "workspace:*"
    },
    "lwc": {
        "modules": [
            {
                "name": "lwc",
                "path": "engine-dom.js"
            },
            {
                "name": "@lwc/synthetic-shadow",
                "path": "synthetic-shadow.js"
            },
            {
                "name": "@lwc/wire-service",
                "path": "wire-service.js"
            }
        ]
    },
    "exports": {
        ".": "./index.js",
        "./aria-reflection": "./aria-reflection.js",
        "./babel-plugin-component": "./babel-plugin-component.js",
        "./compiler": "./compiler.js",
        "./engine-core": "./engine-core.js",
        "./engine-dom": "./engine-dom.js",
        "./engine-server": "./engine-server.js",
        "./errors": "./errors.js",
        "./features": "./features.js",
        "./types": "./types.d.ts",
        "./module-resolver": "./module-resolver.js",
        "./rollup-plugin": "./rollup-plugin.js",
        "./shared": "./shared.js",
        "./ssr-compiler": "./ssr-compiler.js",
        "./ssr-runtime": "./ssr-runtime.js",
        "./style-compiler": "./style-compiler.js",
        "./synthetic-shadow": "./synthetic-shadow.js",
        "./template-compiler": "./template-compiler.js",
        "./wire-service": "./wire-service.js"
    }
}
