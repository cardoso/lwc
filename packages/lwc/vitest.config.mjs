import { defineProject, mergeConfig } from 'vitest/config';
import sharedConfig from '../../vitest.shared';

export default mergeConfig(
    sharedConfig,
    defineProject({
        test: {
            name: 'lwc',
            environment: 'jsdom',
        },
    })
);
