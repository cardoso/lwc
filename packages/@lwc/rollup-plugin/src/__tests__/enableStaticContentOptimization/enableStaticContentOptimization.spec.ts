/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
import path from 'node:path';
import { describe, it } from 'vitest';
import { rolldown } from 'rolldown';
import lwc, { RollupLwcOptions } from '../../index';

describe('enableStaticContentOptimization:', () => {
    async function runRollup(pathname: string, options: RollupLwcOptions) {
        const warnings: any[] = [];

        try {
            const bundle = await rolldown({
                input: path.resolve(__dirname, pathname),
                plugins: [lwc(options)],
                external: ['lwc'],
                shimMissingExports: true,
                onwarn(warning) {
                    warnings.push(warning);
                },
            });

            try {
                const output = await bundle.generate({
                    format: 'esm',
                });

                await bundle.close();

                return {
                    warnings,
                    bundle,
                    output,
                };
            } catch (error) {
                return {
                    warnings,
                    bundle,
                    error,
                };
            }
        } catch (error) {
            return {
                warnings,
                error,
            };
        }
    }

    const configs = [
        {
            name: 'undefined',
            opts: { enableStaticContentOptimization: undefined },
        },
        { name: 'false', opts: { enableStaticContentOptimization: false } },
        { name: 'true', opts: { enableStaticContentOptimization: true } },
        { name: 'unspecified', opts: {}, expected: true },
    ];

    it.concurrent.for(configs)('$name', async ({ opts }, { expect }) => {
        await expect(runRollup('fixtures/basic/basic.js', opts)).resolves.toMatchSnapshot();
    });
});
