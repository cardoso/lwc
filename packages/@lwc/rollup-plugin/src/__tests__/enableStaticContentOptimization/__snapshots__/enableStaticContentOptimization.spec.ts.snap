// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`enableStaticContentOptimization: > 'false' 1`] = `
{
  "bundle": RolldownBuild {},
  "output": {
    "output": [
      {
        "code": "import { LightningElement, freezeTemplate, registerComponent as _registerComponent, registerTemplate } from "lwc";

//#region @lwc/resources/empty_css.css
var default$1 = void 0;

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html
const stc0 = {
	attrs: {
		"src": "http://example.com/img.png",
		"crossorigin": "anonymous"
	},
	key: 0
};
function tmpl($api, $cmp, $slotset, $ctx) {
	const { h: api_element } = $api;
	return [api_element("img", stc0)];
}
var basic_default$1 = registerTemplate(tmpl);
tmpl.stylesheets = [];
tmpl.stylesheetToken = "lwc-5cogeta2bli";
tmpl.legacyStylesheetToken = "fixtures-basic_basic";
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
freezeTemplate(tmpl);

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js
const __lwc_component_class_internal = _registerComponent(class extends LightningElement {}, {
	tmpl: basic_default$1,
	sel: "fixtures-basic",
	apiVersion: 63
});
var basic_default = __lwc_component_class_internal;

//#endregion
export { basic_default as default };",
        "dynamicImports": [],
        "exports": [
          "default",
        ],
        "facadeModuleId": "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        "fileName": "basic.js",
        "imports": [],
        "isDynamicEntry": false,
        "isEntry": true,
        "map": null,
        "moduleIds": [
          "@lwc/resources/empty_css.css",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        ],
        "modules": {
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html": {},
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js": {},
          "@lwc/resources/empty_css.css": {},
        },
        "name": "basic",
        "preliminaryFileName": "basic.js",
        "sourcemapFileName": null,
        "type": "chunk",
      },
      {
        "fileName": "basic.css",
        "name": undefined,
        "originalFileName": null,
        "source": "export default undefined
",
        "type": "asset",
      },
    ],
  },
  "warnings": [],
}
`;

exports[`enableStaticContentOptimization: > 'true' 1`] = `
{
  "bundle": RolldownBuild {},
  "output": {
    "output": [
      {
        "code": "import { LightningElement, freezeTemplate, parseFragment, registerComponent as _registerComponent, registerTemplate } from "lwc";

//#region @lwc/resources/empty_css.css
var default$1 = void 0;

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html
const $fragment1 = parseFragment\`<img src="http://example.com/img.png" crossorigin="anonymous"\${3}>\`;
function tmpl($api, $cmp, $slotset, $ctx) {
	const { st: api_static_fragment } = $api;
	return [api_static_fragment($fragment1, 1)];
}
var basic_default$1 = registerTemplate(tmpl);
tmpl.stylesheets = [];
tmpl.stylesheetToken = "lwc-5cogeta2bli";
tmpl.legacyStylesheetToken = "fixtures-basic_basic";
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
freezeTemplate(tmpl);

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js
const __lwc_component_class_internal = _registerComponent(class extends LightningElement {}, {
	tmpl: basic_default$1,
	sel: "fixtures-basic",
	apiVersion: 63
});
var basic_default = __lwc_component_class_internal;

//#endregion
export { basic_default as default };",
        "dynamicImports": [],
        "exports": [
          "default",
        ],
        "facadeModuleId": "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        "fileName": "basic.js",
        "imports": [],
        "isDynamicEntry": false,
        "isEntry": true,
        "map": null,
        "moduleIds": [
          "@lwc/resources/empty_css.css",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        ],
        "modules": {
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html": {},
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js": {},
          "@lwc/resources/empty_css.css": {},
        },
        "name": "basic",
        "preliminaryFileName": "basic.js",
        "sourcemapFileName": null,
        "type": "chunk",
      },
      {
        "fileName": "basic.css",
        "name": undefined,
        "originalFileName": null,
        "source": "export default undefined
",
        "type": "asset",
      },
    ],
  },
  "warnings": [],
}
`;

exports[`enableStaticContentOptimization: > 'undefined' 1`] = `
{
  "bundle": RolldownBuild {},
  "output": {
    "output": [
      {
        "code": "import { LightningElement, freezeTemplate, registerComponent as _registerComponent, registerTemplate } from "lwc";

//#region @lwc/resources/empty_css.css
var default$1 = void 0;

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html
const stc0 = {
	attrs: {
		"src": "http://example.com/img.png",
		"crossorigin": "anonymous"
	},
	key: 0
};
function tmpl($api, $cmp, $slotset, $ctx) {
	const { h: api_element } = $api;
	return [api_element("img", stc0)];
}
var basic_default$1 = registerTemplate(tmpl);
tmpl.stylesheets = [];
tmpl.stylesheetToken = "lwc-5cogeta2bli";
tmpl.legacyStylesheetToken = "fixtures-basic_basic";
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
freezeTemplate(tmpl);

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js
const __lwc_component_class_internal = _registerComponent(class extends LightningElement {}, {
	tmpl: basic_default$1,
	sel: "fixtures-basic",
	apiVersion: 63
});
var basic_default = __lwc_component_class_internal;

//#endregion
export { basic_default as default };",
        "dynamicImports": [],
        "exports": [
          "default",
        ],
        "facadeModuleId": "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        "fileName": "basic.js",
        "imports": [],
        "isDynamicEntry": false,
        "isEntry": true,
        "map": null,
        "moduleIds": [
          "@lwc/resources/empty_css.css",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        ],
        "modules": {
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html": {},
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js": {},
          "@lwc/resources/empty_css.css": {},
        },
        "name": "basic",
        "preliminaryFileName": "basic.js",
        "sourcemapFileName": null,
        "type": "chunk",
      },
      {
        "fileName": "basic.css",
        "name": undefined,
        "originalFileName": null,
        "source": "export default undefined
",
        "type": "asset",
      },
    ],
  },
  "warnings": [],
}
`;

exports[`enableStaticContentOptimization: > 'unspecified' 1`] = `
{
  "bundle": RolldownBuild {},
  "output": {
    "output": [
      {
        "code": "import { LightningElement, freezeTemplate, parseFragment, registerComponent as _registerComponent, registerTemplate } from "lwc";

//#region @lwc/resources/empty_css.css
var default$1 = void 0;

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html
const $fragment1 = parseFragment\`<img src="http://example.com/img.png" crossorigin="anonymous"\${3}>\`;
function tmpl($api, $cmp, $slotset, $ctx) {
	const { st: api_static_fragment } = $api;
	return [api_static_fragment($fragment1, 1)];
}
var basic_default$1 = registerTemplate(tmpl);
tmpl.stylesheets = [];
tmpl.stylesheetToken = "lwc-5cogeta2bli";
tmpl.legacyStylesheetToken = "fixtures-basic_basic";
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
if (default$1) tmpl.stylesheets.push.apply(tmpl.stylesheets, default$1);
freezeTemplate(tmpl);

//#endregion
//#region packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js
const __lwc_component_class_internal = _registerComponent(class extends LightningElement {}, {
	tmpl: basic_default$1,
	sel: "fixtures-basic",
	apiVersion: 63
});
var basic_default = __lwc_component_class_internal;

//#endregion
export { basic_default as default };",
        "dynamicImports": [],
        "exports": [
          "default",
        ],
        "facadeModuleId": "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        "fileName": "basic.js",
        "imports": [],
        "isDynamicEntry": false,
        "isEntry": true,
        "map": null,
        "moduleIds": [
          "@lwc/resources/empty_css.css",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html",
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js",
        ],
        "modules": {
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.html": {},
          "/Users/cardoso/Documents/salesforce/lwc/packages/@lwc/rollup-plugin/src/__tests__/enableStaticContentOptimization/fixtures/basic/basic.js": {},
          "@lwc/resources/empty_css.css": {},
        },
        "name": "basic",
        "preliminaryFileName": "basic.js",
        "sourcemapFileName": null,
        "type": "chunk",
      },
      {
        "fileName": "basic.css",
        "name": undefined,
        "originalFileName": null,
        "source": "export default undefined
",
        "type": "asset",
      },
    ],
  },
  "warnings": [],
}
`;
